packages/components/src/dateInput/DateInput.js

```diff
@@ -2,58 +2,90 @@
 
 import React from 'react';
 import { defaultTokens } from '@kiwicom/orbit-design-tokens';
-import { format, setMonth } from 'date-fns';
 import { View } from 'react-native';
 import {
   StyleSheet,
   Picker,
   Text,
   TextInput,
 } from '@kiwicom/universal-components';
+import { setYear, setMonth, setDate, format } from 'date-fns';
 
 type Props = {|
-  +onDateChange: (string, string) => void,
-  +date: {
-    +day: string,
-    +month: string,
-    +year: string,
-  },
-  +errors: {
-    +day: string,
-    +year: string,
-  },
+  +onDateChange: Date => void,
+  +date: ?Date,
 |};
 
 type Month = {|
   value: string,
   label: string,
 |};
 
-class DateInput extends React.Component<Props> {
-  getMonthOptions = () => {
-    return [...Array(12).keys()].map<Month>(
-      (currentMonth: number): Month => ({
-        value: String(currentMonth),
-        label: format(setMonth(new Date(), currentMonth), 'MMMM'),
-      }),
-    );
+type State = {|
+  day: string,
+  month: string,
+  year: string,
+|};
+
+const getMonthOptions = () => {
+  return [...Array(12).keys()].map<Month>(
+    (currentMonth: number): Month => ({
+      value: String(currentMonth),
+      label: format(setMonth(new Date(), currentMonth), 'MMMM'),
+    }),
+  );
+};
+
+export default class DateInput extends React.Component<Props, State> {
+  state = {
+    day: '',
+    month: '0',
+    year: '',
+  };
+
+  componentDidUpdate(prevProps: Props, prevState: State) {
+    if (this.state !== prevState) {
+      if (Object.values(this.state).every(el => el !== '')) {
+        if (this.state.year.length === 4) {
+          const isValidDate = this.handleDateValidation();
+          if (isValidDate) this.props.onDateChange(isValidDate);
+        }
+      }
+    }
+  }
+
+  handleDateValidation = () => {
+    const day = parseInt(this.state.day, 10);
+    const month = parseInt(this.state.month, 10);
+    const year = parseInt(this.state.year, 10);
+
+    let date = new Date();
+
+    if (!day) return false;
+    if (!year) return false;
+
+    date = setYear(date, year);
+    date = setMonth(date, month);
+    date = setDate(date, day);
+    return date;
   };
 
-  handleDayChange = (day: string) => {
-    this.props.onDateChange(day, 'day');
+  handleDayChange = (input: string) => {
+    const day = input.slice(0, 2);
+    this.setState({ day });
   };
 
   handleMonthChange = (month: ?string) => {
-    if (month) {
-      this.props.onDateChange(month, 'month');
-    }
+    if (month) this.setState({ month });
   };
 
-  handleYearChange = (year: string) => {
-    this.props.onDateChange(year, 'year');
+  handleYearChange = (input: string) => {
+    const year = input.slice(0, 4);
```

Comment:
Why do you use this?

Reply:
Why are we also cropping the year by doing `slice` (and also checking somewhere else that the length is indeed 4) ? can it be more than 4 digits? I think this should be handled by the component itself (not allow putting more than 4 digits in the year or precisely 4), but I don't know much how it works.

Reply:
Input type number is not handling maxLength as normal input, so this is basically a workaround. I couldnt find any other solution so if you have anything in mind please share :)